File: ff4j-spring-boot-starter/src/main/java/org/ff4j/spring/boot/services/FeatureStoreServices.java
package org.ff4j.spring.boot.services;


import org.apache.commons.lang3.StringUtils;
import org.ff4j.FF4j;
import org.ff4j.cache.FF4jCacheProxy;
import org.ff4j.core.Feature;
import org.ff4j.spring.boot.domain.CacheApiBean;
import org.ff4j.spring.boot.domain.FeatureApiBean;
import org.ff4j.spring.boot.domain.FeatureStoreApiBean;
import org.ff4j.spring.boot.exceptions.FeatureStoreNotCached;
import org.ff4j.spring.boot.domain.GroupDescApiBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.util.CollectionUtils;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Created by Paul
 *
 * @author <a href="mailto:paul58914080@gmail.com">Paul Williams</a>
 */
@Service
public class FeatureStoreServices {
    @Autowired
    private FF4j ff4j;

    public FeatureStoreApiBean getFeatureStore() {
        return new FeatureStoreApiBean(ff4j.getFeatureStore());
    }

    public Collection<FeatureApiBean> getAllFeatures() {
        List<FeatureApiBean> features;
        Map<String, Feature> featureMap = ff4j.getFeatureStore().readAll();
        if (CollectionUtils.isEmpty(featureMap)) {
            features = new ArrayList<>(0);
        } else {
            features = new ArrayList<>(featureMap.size());
            features.addAll(featureMap.values().stream().map(FeatureApiBean::new).collect(Collectors.toList()));
        }
        return features;
    }

    public Collection<GroupDescApiBean> getAllGroups() {
        Map<String, GroupDescApiBean> groups = new HashMap<>();
        Map<String, Feature> featureMap = ff4j.getFeatureStore().readAll();
        if (!CollectionUtils.isEmpty(featureMap)) {
            for (Feature feature : featureMap.values()) {
                initGroupMap(groups, feature.getUid(), feature.getGroup());
            }
        }
        return groups.values();
    }

    private void initGroupMap(Map<String, GroupDescApiBean> groups, String featureUID, String groupName) {
        if (StringUtils.isNotBlank(groupName)) {
            if (groups.containsKey(groupName)) {
                groups.get(groupName).getFeatures().add(featureUID);
            } else {
                groups.put(groupName, new GroupDescApiBean(groupName, new ArrayList<>()));
                groups.get(groupName).getFeatures().add(featureUID);
            }
        }
    }

    public void deleteAllFeatures() {
        ff4j.getFeatureStore().clear();
    }

    public CacheApiBean getFeaturesFromCache() {
        if (ff4j.getFeatureStore() instanceof FF4jCacheProxy) {
            return new CacheApiBean(ff4j.getFeatureStore());
        } else {
            throw new FeatureStoreNotCached();
        }
    }

    public void clearCachedFeatureStore() {
        if (ff4j.getFeatureStore() instanceof FF4jCacheProxy) {
            ((FF4jCacheProxy) ff4j.getFeatureStore()).getCacheManager().clearFeatures();
        } else {
            throw new FeatureStoreNotCached();
        }
    }
}

File: ff4j-spring-services/src/main/kotlin/org/ff4j/services/FeatureStoreServices.kt
/*-
 * #%L
 * ff4j-spring-services
 * %%
 * Copyright (C) 2013 - 2022 FF4J
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package org.ff4j.services

import org.apache.commons.lang3.StringUtils
import org.ff4j.FF4j
import org.ff4j.services.domain.CacheApiBean
import org.ff4j.services.domain.FeatureApiBean
import org.ff4j.services.domain.FeatureStoreApiBean
import org.ff4j.services.domain.GroupDescApiBean
import org.ff4j.services.exceptions.FeatureStoreNotCached
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.util.CollectionUtils
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.util.stream.Collectors

/**
 * Created by Paul
 *
 * @author <a href="mailto:paul58914080@gmail.com">Paul Williams</a>
 */
@Service
class FeatureStoreServices(@Autowired val fF4j: FF4j) {
  fun getFeatureStore(): Mono<FeatureStoreApiBean> {
    return Mono.just(FeatureStoreApiBean(this.fF4j.featureStore))
  }

  fun getAllFeatures(): Flux<FeatureApiBean> {
    val allFeatures = fF4j.featureStore.readAll()
    return if (CollectionUtils.isEmpty(allFeatures)) {
      Flux.empty()
    } else {
      val features = ArrayList<FeatureApiBean>(allFeatures.size)
      features.addAll(allFeatures.values.stream().map { FeatureApiBean(it) }.collect(Collectors.toList()))
      Flux.fromIterable(features)
    }
  }

  fun getAllGroups(): Flux<GroupDescApiBean> {
    val groups = HashMap<String, GroupDescApiBean>()
    val allFeatures = fF4j.featureStore.readAll()
    allFeatures?.let {
      allFeatures.values.forEach {
        initGroup(groups, it.uid, it.group)
      }
    }
    return Flux.fromIterable(groups.values)
  }

  private fun initGroup(groups: HashMap<String, GroupDescApiBean>, uid: String, groupName: String?) {
    groupName?.let {
      if (StringUtils.isNotBlank(groupName)) {
        if (!groups.containsKey(groupName)) {
          groups[groupName] = GroupDescApiBean(groupName, ArrayList())
        }
        groups[groupName]?.features?.add(uid)
      }
    }
  }

  fun deleteAllFeatures() {
    fF4j.featureStore.clear()
  }

  fun getFeaturesFromCache(): Mono<CacheApiBean> {
    fF4j.cacheProxy ?: throw FeatureStoreNotCached()
    return Mono.just(CacheApiBean(fF4j.featureStore))
  }

  fun clearCachedFeatureStore() { // Fixing #218 : If audit is enabled, cannot clear cache.
    val cacheProxy = fF4j.cacheProxy ?: throw FeatureStoreNotCached()
    cacheProxy.cacheManager.clearFeatures()
  }
}
